pieceio/pieceio.go

```diff
@@ -0,0 +1,84 @@
+package pieceio
+
+import (
+	"context"
+	"fmt"
+	"github.com/filecoin-project/go-fil-components/filestore"
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-ipld-prime"
+	"io"
+	"io/ioutil"
+	"os"
+)
+
+type SectorCalculator interface {
+	// GeneratePieceCommitment takes a PADDED io stream and a total size and generates a commP
+	GeneratePieceCommitment(piece io.Reader, pieceSize uint64) ([]byte, error)
+}
+
+type PadReader interface {
+	// PaddedSize returns the expected size of a piece after it's been padded
+	PaddedSize(size uint64) uint64
+}
+
+type CarIO interface {
+	// WriteCar writes a given payload to a CAR file and into the passed IO stream
+	WriteCar(ctx context.Context, bs ReadStore, payloadCid cid.Cid, selector ipld.Node, w io.Writer) error
+	// LoadCar loads blocks into the a store from a given CAR file
+	LoadCar(bs WriteStore, r io.Reader) (cid.Cid, error)
+}
+
+type pieceIO struct {
+	padReader        PadReader
+	carIO            CarIO
+	sectorCalculator SectorCalculator
+	tempDir          filestore.Path
+}
+
+func NewPieceIO(padReader PadReader, carIO CarIO, sectorCalculator SectorCalculator, tempDir filestore.Path) PieceIO {
```

Comment:
I wonder -- should PieceIO just have access to a Filestore directly and be called a PieceStore? And then we don't have to worry about getting things in and out of the filestore?

Reply:
In that case I'd like add a function to create a temporary file through the filestore.